### PR TITLE
Support for Custom Response Errors

### PR DESCRIPTION
**Pull request by phips28**
- add error responses generator from `additionalErrors` for API Gateway
- validate the `additionalErrors`
- bump version
- added release
- Extra: Added vpc policies template

TODO:
- [ ] transfer [claudia-api-errors](https://github.com/phips28/claudia-api-errors) to @claudiajs

### CODE REVIEW

**Comment by gojko** [comment]
Hi,

diff's gone all crazy because you seem to have reformatted everything. Is there a reason for that?

**Comment by phips28** [comment]
yes I saw this now too. My WebStorm reformats everything with my config.
Should I undo it?

**Comment by gojko** [comment]
it will make it significantly easier for me to validate the changes if I can do a quick diff and not have to read through all the source code. If it's possible to use the original formatting with your changes, that would be best. Did you run the tests? jshint should have caught the change in formatting, I'm surprised it allowed you to do it.

**Comment by gojko** [comment]
also, I don't see any new or changed test cases. Can you please add the test cases for your changes, so I can also understand that better.

**Comment by fiznool** [comment]
You can use https://github.com/claudiajs/claudia/pull/69/files?w=0 to view the diff ignoring the whitespace changes.

**Comment by phips28** [comment]
- [x] undo reformats

**Comment by phips28** [comment]
How do I write tests in there?

**Comment by fiznool** [comment]
@phips28 did you see my last comment in the gist?

Using ES6 static methods on the error objects might be a better approach, because then you only need to do the following in the `additionalErrors` array:

``` js
additionalErrors: [
  ApiErrors.BadRequest
]
```

Notice that for convenience only the error class is passed in to the array. Since we'll always be calling `Error.toConfig()` we might as well do this inside the claudia core code:

``` js
additionalErrors().forEach(function (additionalError) {
  var config = additionalError.config();
  results.push({
    code: String(config.code),
    pattern: config.pattern,
    template: config.template,
    headers: headers('error')
  });
});
```

**Comment by phips28** [comment]
@fiznool thx, changed it! much better =)

File: json-templates/vpc-access.json [hunk 1]
```diff
@@ -0,0 +1,79 @@
+{
```
Comment (gojko) [inline json-templates/vpc-access.json]: why is this file in the package? I don't see it being used anywhere?

File: json-templates/vpc-create.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (gojko) [inline json-templates/vpc-create.json]: where is this file used?

File: src/tasks/validate-package.js [hunk 3]
```diff
@@ -46,6 +47,9 @@ module.exports = function validatePackage(dir, functionHandler, restApiModule) {
 						throw routeMessage + 'requests custom headers but does not enumerate any headers';
 					}
 				}
+				if (methodConfig.success && !validHttpCode(methodConfig.success)) {
```
Comment (gojko) [inline src/tasks/validate-package.js]: this is strange, it will fail on `success: { contentType: 'text/plain' }`. Why did you need this?

File: src/tasks/validate-package.js [hunk 4]
```diff
@@ -54,6 +58,25 @@ module.exports = function validatePackage(dir, functionHandler, restApiModule) {
 						throw routeMessage + 'error template requests custom headers but does not provide defaults';
 					}
 				}
+				if (methodConfig.error && methodConfig.error.defaultCode && !validHttpCode(methodConfig.error.defaultCode)) {
```
Comment (gojko) [inline src/tasks/validate-package.js]: I think this should just be error.code, not error.defaultCode, to keep things backwards compatible

File: src/tasks/validate-package.js [hunk 5]
```diff
@@ -54,6 +58,25 @@ module.exports = function validatePackage(dir, functionHandler, restApiModule) {
...
+					throw routeMessage + 'error defaultCode ' + methodConfig.error.defaultCode + ' is no valid http code';
+				}
+				if (methodConfig.error && methodConfig.error.additionalErrors) {
```
Comment (gojko) [inline src/tasks/validate-package.js]: please also add tests for this to /spec/validate-package-spec.js

File: json-templates/vpc-access.json (hunk 1 above)
Comment (phips28) [inline json-templates/vpc-access.json]: I just added this file to have an example for VPC policy. Not necessary for this feature.

File: json-templates/vpc-create.json (hunk 2 above)
Comment (phips28) [inline json-templates/vpc-create.json]: I just added this file to have an example for VPC policy. Not necessary for this feature.

Comment (gojko) [inline json-templates/vpc-create.json]: ok, I guess we should remove it from this pull request

File: src/tasks/validate-package.js (hunk 3 above)
Comment (phips28) [inline src/tasks/validate-package.js]: removed it

File: src/tasks/validate-package.js (hunk 4 above)
Comment (phips28) [inline src/tasks/validate-package.js]: you can still use error.code
should I remove the defaultCode again?

Comment (gojko) [inline src/tasks/validate-package.js]: I don't know, I didn't really understand why it is there. That's why I asked. Is there a special reason why you think people should use that? otherwise, it's best to keep it simple, so `code` is the default, and additional errors are extras

**Comment by gojko** [comment]
@phips28 >How do I write tests in there?

specs are in the spec folder, there is a ton of examples already. You will likely have to change the rebuild-web-api-spec.js to add extra tests for the error codes, including the variants of the following tests when additional error codes are specified:

https://github.com/claudiajs/claudia/blob/4da78e7755084602dbd7f4d0de17404294d5927e/spec/rebuild-web-api-spec.js#L728

https://github.com/claudiajs/claudia/blob/4da78e7755084602dbd7f4d0de17404294d5927e/spec/rebuild-web-api-spec.js#L905

https://github.com/claudiajs/claudia/blob/4da78e7755084602dbd7f4d0de17404294d5927e/spec/rebuild-web-api-spec.js#L941

this whole subsection probably needs revisiting:
https://github.com/claudiajs/claudia/blob/4da78e7755084602dbd7f4d0de17404294d5927e/spec/rebuild-web-api-spec.js#L1098

https://github.com/claudiajs/claudia/blob/master/contributing.md explains how to run the tests

**Comment by gojko** [comment]
@fiznool 
> Using ES6 static methods on

Until Lambda phases out 0.10 (in october), any core piece of claudia has to work with 0.10 as well, so ES6 isn't an option

File: src/tasks/validate-package.js (hunk 4 above)
Comment (phips28) [inline src/tasks/validate-package.js]: true, removed defaultCode.
It was in @fiznool's gist, and I copy paste it

File: src/tasks/rebuild-web-api.js [hunk 6]
```diff
@@ -295,10 +295,11 @@ module.exports = function rebuildWebApi(functionName, functionVersion, restApiId
 				}
 				if (additionalErrors()) {
 					additionalErrors().forEach(function (additionalError) {
+						var additionalErrorConfig = additionalError.toConfig();
```
Comment (gojko) [inline src/tasks/rebuild-web-api.js]: This would assume that we can only use the special config errors with this, and that's different from what we planned to do. The whole point of extracting errors into a separate package was so that people can just configure a few errors by hand if they want, and if someone wants to use generic utility handlers, they can.

until now, the config was just data, not functions. I'd try to keep that the same. I understand why you didn't like calling the .toConfig all the time, but perhaps we can find a different solution for that part in claudia-api-errors package.

File: src/tasks/validate-package.js (hunk 5 above)
Comment (phips28) [inline src/tasks/validate-package.js]: done, check it out if it is ok

File: src/tasks/rebuild-web-api.js (hunk 6 above)
Comment (phips28) [inline src/tasks/rebuild-web-api.js]: you can override the Errors https://github.com/phips28/claudia-api-errors#override-error-classes
There you can set a different output template.
what else would you like to configure for an error?
---
so we should call the `toConfig()` function directly for every function to make the chaching work?
```
additionalErrors: [
  new ApiErrors.BadRequest().toConfig(),
],
```

**Comment by phips28** [comment]
@gojko 

> ES6 isn't an option

Then wait till october ;) I have no rush merging it.
I am using my own repos for now in our API.

**Comment by fiznool** [comment]
The way I originally saw this working is that we'd use 2 separate properties for the errors, that you would use in the third argument passed to the handler function.

The following would be used for the 'regex' style error handling. This is the support we discussed that would be built in to the claudia core. Notice that the property name is `additionalCodes`.

``` js
error: {
  additionalCodes: [{
    code: 400,
    pattern: '^Bad Request:.*',
    template: '$input.path(\'$.errorMessage\')'  // Optional, defaults to the value shown here
  }]
}
```

The following would be supported for the `ApiError` objects. Notice that the property name is `additionalErrors`.

``` js
error: {
  additionalErrors: [
    ApiErrors.BadRequestError
  ]
}
```

I thought that in claudia, we could then treat the `additionalCodes` and `additionalErrors` separately:

- If there are any `additionalCodes`, use the `code`, `pattern` and `template` directly to configure API Gateway.
- If there are any `additionalErrors`, assume this is an error object and call `toConfig()` on the object to get the `code`, `pattern` and `template` properties needed to configure API Gateway.

If its undesirable to be parsing `additionalErrors` in claudia, we could use the following instead for the `ApiErrors`:

``` js
error: {
  additionalCodes: [
    ApiErrors.BadRequestError.toConfig()
  ]
}
```

So we would instead use `additionalCodes` and explicitly call `toConfig()`. However I figured that since we'd be calling `toConfig()` on every object, it made more sense to call this from within claudia.

My apologies for not making all of this clearer earlier, I can see how the confusion occurred.

**Comment by fiznool** [comment]
@gojko Regarding the ES6 support, I think we have 2 options:

- Limit the `claudia-api-errors` to users that are using node v4. It's an opt-in module and so this limitation will probably be ok IMHO, and besides as you mention, AWS Lambda is only supporting non-ES6 environments for another month.
- Use babel to transpile the ES6 code down to ES5 to support node v0.10.

My own preference is the first option, since we only need to wait a month until we can use ES6 natively across all supported lambda functions.

**Comment by phips28** [comment]
@fiznool I had `additionalCodes` in my first version, worked perfectly.
Then I moved on to add the ApiErrors class with `additionalErrors` - and removed the `additionalCodes` 🙈 - I thought I don't need it anymore when I have the error classes.

I don't like the direct JSON config with `additionalCodes`. Who one is returning non-json errors from their API? If you want to return a json error object, then you have write a lot of config with the stringify, code pattern and template. It's much easier to just use the ApiErrors

Do you really want plain json support?

**Comment by fiznool** [comment]
> I don't like the direct JSON config with additionalCodes. Who one is returning non-json errors from their API? If you want to return a json error object, then you have write a lot of config with the stringify, code pattern and template. It's much easier to just use the ApiErrors

> Do you really want plain json support?

I agree with you actually, but I discussed this with @gojko on gitter and he prefers to support `additionalCodes` with plain config first, and with the optional `ApiErrors` support as a separate module. You can see the relevant discussion [here](https://gitter.im/claudiajs/claudia?at=57cfd23a77c1b70d7ff08a1c). So I guess we need to support both :smile:

**Comment by phips28** [comment]
ok `additionalCodes` would be easy to add, but it makes no sense IMO.
Should I add it @gojko?

**Comment by gojko** [comment]
sorry, I'm not sure I fully understand your proposed difference between additionalCodes and additionalErrors. Maybe you can provide a few examples of how the config would look like and we can discuss?

My point was that configuration has to be plain data. It can't contain functions. Functions don't go into serialization nicely, which means it's impossible to take a SHA signature and check if the config changed easily. Allowing functions in config makes it very difficult to validate the configuration without executing the package, and I don't want Claudia to invoke random functions which might have side-effects (because they are written by other people, not just us), during packaging.

**Comment by phips28** [comment]
The API Gateway cache works fine with the functions:
```javascript
additionalErrors: [
        ApiErrors.NotFound,
        ApiErrors.Conflict,
        ApiErrors.BadRequest,
      ],
```
When I remove or add a new error it rebuilds the IntegrationResponses.

Or do you mean another cache?

**Comment by gojko** [comment]
@phips28 yes, remove or add. how about change the function body? APi config cache will have no idea about that.

**Comment by phips28** [comment]
ah true. so the easiest way for now would be to call the toConfig() directly on every route setup.
```
additionalErrors: [
        new ApiErrors.NotFound().toConfig(),
      ],
```
and check in `validatePackage` if the additonalErrors are just JSON objects

**Comment by gojko** [comment]
I think that would be good. We can work on cleaning up the syntax for ApiErrors later, and figure out a way how to make that simpler, so it doesn't have to invoke a function call (for example, pre-cache the config somewhere in the module under a similar name)

**Comment by gojko** [comment]
that's why I suggested just focusing on this part of the puzzle first, once we get the config syntax into a shape that everyone likes, we'll know how to do the opt-in module so it speeds it up

**Comment by phips28** [comment]
What do you think about this idea:

### Create config templates:
file: **api-errors-config.js** (could be in an separate module like the _claudia-api-errors_, or user can just copy it in there projects)
```javascript
'use scrict';

const errorTemplateDynamic = '#set ($errorMessageObj = $util.parseJson($input.path(\'$.errorMessage\')).data)' +
  '{' +
  '#foreach($key in $errorMessageObj.keySet())' + // iterate over each error element
  '"$key":"$errorMessageObj.get($key)"' + // key: value
  '#if($foreach.hasNext),#end' + // add a comma if hasNext element
  '#end' +
  '}';

module.exports = {
  BAD_REQUEST: {
    code: 400,
    pattern: '^\\{\\"code\\":400.*',
    template: errorTemplateDynamic
  },
  UNAUTHORIZED: {
    code: 401,
    pattern: '^\\{\\"code\\":401.*',
    template: errorTemplateDynamic
  },
  // add all other error codes here...
};
```

### Configure route with _existing_ templates:
```javascript
const API_ERRORS_CONFIG = require('api-errors-config.js');

error: {
      code: 500, // optional, default: 500
      additionalErrors: [
        API_ERRORS_CONFIG.BAD_REQUEST,
        API_ERRORS_CONFIG.CONFLICT,
      ],
    },
```

### Configure route with _custom_ templates:
```javascript
const API_ERRORS_CONFIG = require('api-errors-config.js');

const CUSTOM_ERROR = {
    code: 400,
    pattern: '^Bad Request:.*',
    template: '$input.path(\'$.errorMessage\')'  // Optional, defaults to the value shown here
  };

error: {
      code: 500, // optional, default: 500
      additionalErrors: [
        API_ERRORS_CONFIG.BAD_REQUEST,
        API_ERRORS_CONFIG.CONFLICT,
        CUSTOM_ERROR,
      ],
    },
```

### validatePackage
In claudia.validatePackage you can now easily check:
- additionalErrors must be an array
- elements in array must be an object
- object in array must have **code** (int||string), **pattern** (check if regex is valid), **template** (check validation => [apache velocity](http://velocity.apache.org/engine/devel/user-guide.html#user-guide-contents))

### Throw errors
```javascript
throw new Error(`{"code":${400},"data":${JSON.stringify({ message: 'this is a very bad request' })}}`);
throw new Error(`{"code":${API_ERRORS_CONFIG.BAD_REQUEST.code},"data":${JSON.stringify({ message: 'this is a very bad request' })}}`);
// custom error:
throw new Error('Bad Request: this is a very bad request');
```

## 2nd Step
Create a helper function to throw the error without dealing with this `stringify` when using _existing_ errors. something like:
```javascript
function getError(code, data) {
    if (!code) {
        code = 500; // INTERNAL SERVER ERROR
    }
    if (!data) {
      data = 'Something went wrong';
    }
    if (typeof data !== 'object') {
      data = { message: data };
    }
    const serializedData = JSON.stringify(data);
    return `{"code":${code},"data":${serializedData}}`;
  }
```

### Throw errors nr.2
```javascript
throw new Error(getError(400, 'this is a very bad request'));
throw new Error(getError(API_ERRORS_CONFIG.BAD_REQUEST.code, 'this is a very bad request'));
```
Its not as beautiful as with ES6 classes. But we can do this with Error classes of course.
Just need to figure out a good file/module/exports structure.

---

What do you guys think? Its fully customizable and JSON driven. 
Feedback & improvements welcome 😎

**Comment by gojko** [comment]
if I read this correctly, the full api config, with all the optional fields filled in, would look like:

```js
get('/', function (request) {...}, { 
  error: {
    code: 500, /* default code, for empty template */
    contentType: 'application/json', 
    headers: { 'X-Api-Error': 'Unknown' } 
    additionalErrors: [{
      code: 400,
      pattern: '^Bad Request:.*',
      template: '$input.path(\'$.errorMessage\')'
    }]
  }
});
```

If so, I like it. We should also support `contentType` and `headers` in additionalErrors to make it fully consistent with the default error, and probably use the default content type and headers as defaults (so they can be overwritten by additionalerrors, but if not set, the global ones are used. )

Btw, the default template really depends on the content type, so it's dangerous to assume it will default to '$input.path(\'$.errorMessage\')'. So far, if the clients used `application/json` as the error content type, the template would just be pass-through (so api gateway would serialize the entire error). If the content type wasn't `application/json`, then it would just pull out the message. Check out this piece of code https://github.com/claudiajs/claudia/blob/137a86e13f84b5a2c4c3a3cbf309c558ec58ce78/src/tasks/rebuild-web-api.js#L189

**Comment by gojko** [comment]
another thing this raises is whether the default error should also have an optional template to make it consistent with the rest. so eg:
1. if the template is not set on default error, the current system is used (json->complete serialization, anything else->message text). 
2. If the template is set, the template gets used. 
3. If the template is not set on an additional error, and the content type is not set as well, the default error template is used, according to the rules above. 
4. if the template is not set on an additional error, but the content type is set, then the default template isn't used. instead, it's calculated according to rule #1 above. 
5. if the template is set on the additional error, it is used instead. 

how does that sound to you?

**Comment by phips28** [comment]
Hi @gojko sounds good so far. Sry was an a short trip and had no time to reply.
today I worked again on the errors.
checkout `claudia-api-errors` -> https://github.com/phips28/claudia-api-errors/tree/2016-09-17_only_const_json_construct & `claudia` -> master

**Comment by gojko** [comment]
ok, this is shaping up very nicely. I should have a bit of time this week to merge in everything. thanks for persevering with this.

**Comment by gojko** [comment]
@phips28 API gateway just released a major update, with what it seems a generic support for multiple response codes. https://aws.amazon.com/blogs/aws/api-gateway-update-new-features-simplify-api-development/

I need to look into this a bit more, but at least according to the article, it seems that api gw can now  accept an object with payload, headers and response code to return to http. so we may not need the whole complexity with templates and pattern matching.

**Comment by gojko** [comment]
more info here... it seems as a new method integration type, which supports a response in a specific structure: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-set-up-simple-proxy.html#api-gateway-simple-proxy-for-lambda-output-format -- I'll start working on integrating that over the weekend, that will be a better systematic solution for multiple response codes

**Comment by fiznool** [comment]
Great news that API Gateway and Lambda integration has been updated. This should make it much easier to return custom errors without all the hackery! Looking forward to claudia supporting this new arrangement :smile:

**Comment by phips28** [comment]
Oooomg! 💩 :( wasted time. But good to see progress on their side.

**Comment by phips28** [comment]
@gojko do you merge the claudia-api-errors too?
Or do you only implement the new APIG changes?

**Comment by gojko** [comment]
@phips28 I'm not sure, didn't have time to properly test new API GW changes yet. If we still need error regex patterns, i'll merge api errors. if regex patterns are no longer needed, then I guess there wouldn't be much point merging that.

**Comment by phips28** [comment]
Yeah thought so too.
Anyways, my claudia-api-errors/master + claudia/master is ready to use. We use it for now internally.

**Comment by gojko** [comment]
@phips28 I've done an initial round of testing the new integration type, and it looks as if it will not require regexes to match anything -- the lambda can just return a numeric response code and headers as it likes. So I'll close this PR now, and start working on the migration to the new API GW features.

**Comment by gojko** [comment]
2.0 now has support for custom response codes. Check out the [docs](https://github.com/claudiajs/claudia-api-builder/blob/master/docs/api.md#dynamic-responses) and an [example project](https://github.com/claudiajs/example-projects/tree/master/web-api-custom-status-code)
